### PR TITLE
OCPEDGE-649: check if CloudCredentialManager present

### PR DESCRIPTION
**Pull request by qJkee**
Add check if CCMO is present within the cluster. If not, then do not pass the flag to kubelet related to the cloud providers

Here I try to read the install config from the cluster or from the disc and detect enabled capabilities.
If CloudController capability is disabled(cloud controller manager operator is not installed to the cluster) we return an empty value passed to the --cloud-provider kubelet flag. We pass external only in case if cloud provider is actually external to support manual installation of providers

This is a new PR in order to keep actual context. The old one is https://github.com/openshift/machine-config-operator/pull/3999

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -37,16 +37,16 @@ require (
 	github.com/spf13/pflag v1.0.6-0.20210604193023-d5e0c0615ace
 	github.com/stretchr/testify v1.8.4
 	github.com/vincent-petithory/dataurl v1.0.0
-	golang.org/x/net v0.18.0
+	golang.org/x/net v0.19.0
```
Comment (cdoern) [inline go.mod:40]: might have asked this, but any reason for these bumps?

File: go.mod [hunk 2]
```diff
@@ -330,3 +330,5 @@ require (
 )
 
 replace k8s.io/kube-openapi => github.com/openshift/kube-openapi v0.0.0-20230816122517-ffc8f001abb0
+
+replace github.com/openshift/api => github.com/qjkee/api v0.0.0-20240209124943-dfcecbd06e01
```
Comment (cdoern) [inline go.mod:334]: again, we would like this merged before we can merge this cc @JoelSpeed

File: pkg/controller/cloud-controller-cap/capability.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	for _, opt := range opts {
+		opt(&conf)
+	}
+
+	// check if running in hypershift and return true because
+	// hypershift does not support capabilities
+	external, err := isTopologyExternal(conf.manifestDir)
+	if err != nil {
+		klog.Infoln("failed to check if topology is external", err)
+	} else if external {
```
Comment (cdoern) [inline pkg/controller/cloud-controller-cap/capability.go:69]: makes sense. just hypershift, right?

File: pkg/controller/cloud-controller-cap/capability.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			cfg.restConfig = restCfg
+		}
+		cl, err := openshiftconfigclientset.NewForConfig(cfg.restConfig)
+		if err != nil {
+			return false, fmt.Errorf("failed to create clientset from rest config, %w", err)
+		}
+		cfg.configClientSet = cl
+	}
+
+	cv, err := cfg.configClientSet.ConfigV1().ClusterVersions().Get(context.Background(), "version", metav1.GetOptions{})
```
Comment (cdoern) [inline pkg/controller/cloud-controller-cap/capability.go:130]: can we use a lister here?

File: pkg/controller/cloud-controller-cap/capability.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+// checkInstallConfigForCCMODisabled reads install config from the filesystem
+// and checks if CloudController capability disabled
+func checkInstallConfigForCCMODisabled(cfg config) (bool, error) {
+	f, err := os.ReadFile(filepath.Join(cfg.manifestDir, installConfigFilename))
+	if err != nil {
+		return false, fmt.Errorf("failed to read cluster config, %w", err)
+	}
+
+	obji, err := runtime.Decode(kscheme.Codecs.UniversalDecoder(corev1.SchemeGroupVersion), f)
```
Comment (cdoern) [inline pkg/controller/cloud-controller-cap/capability.go:156]: interesting. can we not just unmarshal into a configmap and if we fail, error?

**Review by cdoern — APPROVED** [review]
looks good. I caught a few things looking over this briefly. I will give this another look next week. I agree with your overall choices and defaults in most of these scenarios!

File: pkg/controller/cloud-controller-cap/capability.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	disabled, err := checkClusterVersionForCCMODisabled(conf)
+	if err != nil {
+		klog.Errorln("failed to check if CloudController disabled via clusterversion, trying install config", err)
+	} else {
+		return disabled, err
+	}
+
+	disabled, err = checkInstallConfigForCCMODisabled(conf)
```
Comment (JoelSpeed) [inline pkg/controller/cloud-controller-cap/capability.go:80]: The cluster version should be the source of truth, why is this required? If the clusterversion checks fail I would not expect a fallback

File: pkg/controller/template/render.go [hunk 7]
```diff
@@ -375,29 +376,32 @@ func skipMissing(key string) (interface{}, error) {
...
-		case configv1.AWSPlatformType, configv1.AzurePlatformType, configv1.OpenStackPlatformType, configv1.VSpherePlatformType:
-			return strings.ToLower(string(cfg.Infra.Status.PlatformStatus.Type)), nil
-		case configv1.GCPPlatformType:
-			return "gce", nil
-		default:
+	external, err := cloudprovider.IsCloudProviderExternal(cfg.Infra.Status.PlatformStatus, cfg.FeatureGateAccess)
+	if err != nil {
+		klog.Errorln("failed to check if cloud provider external", err)
+	} else if external {
+		if cfg.CloudControllerDisabled && isCCMNone(cfg.Infra.Status.PlatformStatus) {
```
Comment (JoelSpeed) [inline pkg/controller/template/render.go:391]: The latter half of this check is already done in [library-go](https://github.com/openshift/library-go/blob/4602d24d27bc94d0cbee0d4646eee867e808c5b4/pkg/cloudprovider/external.go#L45), it's impossible to reach this code on platforms where the cloud controller manager may be disabled.
